src/dbnode/storage/index/metrics.go

```diff
@@ -0,0 +1,160 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"math"
+	"time"
+
+	"github.com/uber-go/tally"
+
+	"github.com/m3db/m3/src/x/instrument"
+)
+
+// NewQueryMetrics returns a new QueryMetrics.
+func NewQueryMetrics(name string, scope tally.Scope) QueryMetrics {
+	return QueryMetrics{
+		ByRange: NewQueryRangeMetrics(name, scope),
+		ByDocs:  NewCardinalityMetrics(name, scope),
+	}
+}
+
+// QueryMetrics is a composite type of QueryDurationMetrics and QueryCardinalityMetrics.
+type QueryMetrics struct {
+	ByRange QueryDurationMetrics
+	ByDocs  QueryCardinalityMetrics
+}
+
+type queryRangeHist struct {
+	threshold time.Duration
+	timing    tally.Histogram
+}
+
+func newQueryRangeHist(
+	threshold time.Duration,
+	value string,
+	scope tally.Scope,
+) *queryRangeHist {
+	return &queryRangeHist{
+		threshold: threshold,
+		timing: scope.
+			Tagged(map[string]string{"range": value}).
+			Histogram("timing", instrument.SparseHistogramTimerHistogramBuckets()),
+	}
+}
+
+type queryDurationMetrics struct {
+	histograms []*queryRangeHist
+}
+
+func (bm *queryDurationMetrics) Record(
+	queryRange time.Duration,
+	queryRuntime time.Duration,
+) {
+	for _, m := range bm.histograms {
+		if queryRange <= m.threshold {
+			m.timing.RecordDuration(queryRuntime)
+			return
+		}
+	}
+}
+
+// NewQueryRangeMetrics creates query duration metrics.
+func NewQueryRangeMetrics(metricType string, scope tally.Scope) QueryDurationMetrics {
+	scope = scope.SubScope(metricType).SubScope("query_range")
+	return &queryDurationMetrics{
+		histograms: []*queryRangeHist{
+			newQueryRangeHist(time.Minute*5, "5m", scope),
+			newQueryRangeHist(time.Minute*30, "30m", scope),
+			newQueryRangeHist(time.Hour, "1hr", scope),
+			newQueryRangeHist(time.Hour*6, "6hr", scope),
+			newQueryRangeHist(time.Hour*24, "1d", scope),
+			newQueryRangeHist(time.Hour*24*5, "5d", scope),
+			newQueryRangeHist(time.Hour*24*10, "10d", scope),
+			newQueryRangeHist(time.Hour*24*30, "30d", scope),
+			newQueryRangeHist(time.Duration(math.MaxInt32), "max", scope),
+		},
+	}
+}
+
+// QueryDurationMetrics are timing metrics bucketed by the query window (start, end).
+type QueryDurationMetrics interface {
+	// Record records the runtime for queries given their range.
+	Record(queryRange time.Duration, queryRuntime time.Duration)
+}
+
+type cardinalityHist struct {
+	threshold int
+	timing    tally.Histogram
+}
+
+func newCardinalityHist(
+	threshold int,
+	value string,
+	scope tally.Scope,
+) *cardinalityHist {
+	return &cardinalityHist{
+		threshold: threshold,
+		timing: scope.
+			Tagged(map[string]string{"count": value}).
+			Histogram("timing", instrument.SparseHistogramTimerHistogramBuckets()),
+	}
+}
+
+type queryCardinality struct {
+	cardinalityHists []*cardinalityHist
+}
+
+func (bm *queryCardinality) Record(seriesCount int, queryRuntime time.Duration) {
```

Comment:
This should be `docsCount` yeah? Maybe we should rename all this "cardinality" metrics naming to "queryDocsCount" metrics?